### ISSUE TITLE
Plans to merge with helm core

### ISSUE THREAD

**Issue by vishaltak** [databus23/helm-diff#211]
Hello folks,

This is related to [issue](https://github.com/helm/helm/issues/3156) which is basically asking if helm-diff can become part of the helm core ([here](https://github.com/helm/helm/issues/3156#issuecomment-525077167)). The maintainer mentions that this question can be answered by the folks at "helm-diff" because it concerns itself with governance model and so on.

So, are there any plans to get merged with the helm core? It would definitely make the helm'ing experience much better. The possibility of showing "diff" directly through helm commands using flags is promising.

**Comment by mumoshu** [comment]
@vishaltak Hey! I think we don't have any specific "plan". However, I understand that people want it. Also, I'd be happy to keep maintaining it even after it's somehow merged into the upstream.

Maybe the biggest question is - who's motivated enough to drive this matter? Personally - from a helm-diff maintainer's perspective - I'm fine with this being a plugin. Second, the governance issue. Would the helm core team accept us helm-diff maintainers into the core team?

@databus23 Do you have any motivation to merge this with the upstream helm project?

**Comment by vishaltak** [comment]
@mumoshu  - Hey! Thanks for taking this in a positive direction.
@bacongobbler can answer the question related to maintainers position you raised.

While I'm relatively new to helm'ing in general, I would be more than happy to help in something that's cut out for my size.

**Comment by bacongobbler** [comment]
This would be a good question to raise at the Helm developer call. There's one today at 9:30AM PST if the maintainers of helm-diff would like to attend and pitch the idea. I will not be attending today, but that would be a good first step.

**Comment by max-rocket-internet** [comment]
I will add my 2 cents....

Using helm-diff is a game changer. For us, it's become vital to be able to see changes before they are applied, just like in Terraform. By having it in a separate repo as a plugin is kind of making the diff functionality a second class citizen and many people won't use it or even know it exists.

IMO having it merged into the main project would be extremely beneficial to Helm users 🙂

**Comment by max-rocket-internet** [comment]
Not stale. Still relevant. Would love to see some movement here.

**Comment by Chili-Man** [comment]
Agreed with @max-rocket-internet ; the ability to perform a diff a la Terraform plan is extremely important for validating the changes that will be deployed.  Adding it to Helm core would reduce the friction tremendously and allow for safer deployments out of the box.

**Comment by pio2pio** [comment]
We already have helm-diff plugin installed in every pipeline to achieve the plan/apply stages; but this feels like missing functionality of the core software. I am looking forward for this to evolve into the core branch like terragrunt ideas ending up in the core TF.

**Comment by lud97x** [comment]
+1
We use also helm-diff in our pipelines.

**Comment by vishaltak** [comment]
> Not stale. Still relevant. Would love to see some movement here.

☝️

**Comment by kamalmarhubi** [comment]
> Not stale. Still relevant.
:point_up: 

@databus23 could you label this as pinned?

**Comment by mumoshu** [comment]
Although I think this is useful from the user's perspective, it doesn't seem as straightforward or easy as everyone thinks.
There're many options and modes in helm-diff today and I don't think if all of them make sense to be within or harmonize with the helm core.

So definitely it isn't a thing like "just submit a pull request against helm/helm".

If anyone could step up to sort every detail and write a HIP(https://github.com/helm/community/tree/main/hips) for this and drive the project of merging helm-diff into the core, it would be great. But I'm not willing to do it myself without having a considerable amount of support/sponsorship, which is very unlikely to happen.

**Comment by kfox1111** [comment]
Not sure I follow. Are there use cases helm-diff currently provides that you think should be removed?

I think one of the big issues is the limbo state the plugin has because of the copyright issue. If that can be resolved, it may be much more likely that other folks might step in. As, I think it has a chilling effect on contributions. Folks may be waiting and seeing what happens first. If it was moved to helm to manage, that would get resolved at least.

Refactoring/rewriting it so that it would more easily fit into core is a separate issue I think.

I'm willing to help out some. I don't have a lot of time though.

**Comment by mumoshu** [comment]
@kfox1111 I personally think commands other than `helm diff upgrade` is not strictly needed. But that's not important. The problem is more about no one knowing the exact goal of merging.

> I think one of the big issues is the limbo state the plugin has because of the copyright issue

What do you mean by this? I see no copyright issue neither in helm-diff nor helm.

> Folks may be waiting and seeing what happens first. If it was moved to helm to manage, that would get resolved at least.

If you're willing to contribute to an opensource project, all I can say is you should just start it today.

For me, I tend to start by contributing small enhancements and tasks, even chime in to help answering random questions in issues to help maintainers get time to work on bigger things.

> Refactoring/rewriting it so that it would more easily fit into core is a separate issue I think.

Yeah I agree. But again, the problem is more about that no one has stepped up to drive that huge work, perhaps no one knows if worth the effort.

**Comment by max-rocket-internet** [comment]
> I personally think commands other than helm diff upgrade is not strictly needed

100% agree. We really just want to know what will happen before upgrading a chart, just like `terraform plan`.

**Comment by R-omk** [comment]
> I personally think commands other than helm diff upgrade is not strictly needed

Change detection is needed not only before the actual upgrade, we are going to use this mechanism as a background service for monitoring. We believe that skew between helm release manifests and actual state are not the norm.

**Comment by kfox1111** [comment]
> @kfox1111 I personally think commands other than `helm diff upgrade` is not strictly needed. But that's not important. In that regard, the problem is more about no one knowing the exact goal of merging.

Yeah, that could be. I typically only helm diff upgrade.
IE, what is the pipeline thats going to run going to do to the prod system.

Maybe we need a user poll or something asking what features are typically used?

Also defining a goal like you suggest would be useful. Mine has been, Its a pain to deal with the plugin in automated pipelines in particular. I'd like a job using image https://hub.docker.com/r/lachlanevenson/k8s-helm to "Just work(TM)". It working anywhere helm's downloaded without having to fiddle with running arbitrary install scripts from the internet, that would be great too.

Its currently easy to helm upgrade something in an arbitrary pipeline. its a little less easy to helm diff upgrade too to better see whats going to happen.

> > I think one of the big issues is the limbo state the plugin has because of the copyright issue
> 
> What do you mean by this? I see no copyright issue neither in helm-diff nor helm.

Maybe I misspoke a little bit there. Let me try again. There's opensource licensing, and opensource governance. This project is opensource licensed. But the governance is more up in the air. Some developers (and even some users) want to see the governance well defined before contributing (or using) something. (Something like joining the CNCF)

> > Folks may be waiting and seeing what happens first. If it was moved to helm to manage, that would get resolved at least.
> 
> If you're willing to contribute to an opensource project, all I can say is you should just start it today.

The governance is a concern. With it being not well defined, there are kind of two paths moving forward:
1. The governance is moved over to the CNCF. This allows the existing code to be massaged into a form that could be merged directly into helm over time.
2. The code wont receive such a governance change. To move forward with the goal, helm gets a complete re-implementation of helm diff within tree.

This leaves devs to make one of 3 choices:
1. Don't contribute until option 1 or option 2 is picked from above. (I'm kind of in this boat)
2. contribute to helm-diff now, but hope that option 1 will happen. If option 2 happens, its not the best spent effort.
3. contribute to helm re-implementation of helm diff. thats a lot of work...

The easiest thing to do is choose 1 and just wait longer. That I think is what has been happening.

> > Refactoring/rewriting it so that it would more easily fit into core is a separate issue I think.
> 
> Yeah I agree. But again, the problem is more about that no one has stepped up to drive that potentially huge work, perhaps no one knows if worth the effort too.

Yeah. But I think the road block described above is part of it too. The potential effort is even more without knowing the way governance will work.

**Comment by mumoshu** [comment]
> Maybe we need a user poll or something asking what features are typically used?

@kfox1111 Sounds good.

> There's opensource licensing, and opensource governance. This project is opensource licensed. But the governance is more up in the air. Some developers (and even some users) want to see the governance well defined before contributing (or using) something. (Something like joining the CNCF)

I can imagine that some enterprise users do care about it. I don't oppose fixing an issue like that. Either joining CNCF or making it a part of helm core is okay for me. My opinion is, though, one of those enterprise users who cares about it should sponsor and contribute to this project so that it better fits their use-case and relief.

> 1. The governance is moved over to the CNCF. This allows the existing code to be massaged into a form that could be merged directly into helm over time.
> 2. The code wont receive such a governance change. To move forward with the goal, helm gets a complete re-implementation of helm diff within tree.

For me, either is ok. The blocker is still just that no one seems to be going to drive it towards neither the option 1 nor the option 2. I, as a maintainer of helm-diff, am fine with either option. So if you're going to step up to drive either 1 or 2, I'd just say go ahead!

But here's my two cents. Even though helm-diff is awesome and looks very mature project, it's still not big enough to be a part of CNCF, I think. From my imaginary perspective of board members of Helm and CNCF, I'd guess the only option can be 1 in practice. YMMV though.

Anyway, thanks again for your feedback! This turned out to be a valuable discussion to build a foundation towards the future of helm-diff.

**Comment by mumoshu** [comment]
https://github.com/actions-runner-controller/actions-runner-controller/blob/8a73560dbccc46df9fda7284173057b556b909fc/controllers/runner_controller.go#L571-L572

L571 is the part that records the event and L572 is for the error message that should have been recorded in your controller-manager log. You should be able to look for messages emitted by L572 to analyze the cause of the issue.

**Comment by kfox1111** [comment]
> For me, either is ok. The blocker is still just that no one seems to be going to drive it towards neither the option 1 nor the option 2. I, as a maintainer of helm-diff, am fine with either option. So if you're going to step up to drive either 1 or 2, I'd just say go ahead!
> 
> But here's my two cents. Even though helm-diff is awesome and looks very mature project, it's still not big enough to be a part of CNCF, I think. From my imaginary perspective of board members of Helm and CNCF, I'd guess the only option can be 1 in practice. YMMV though.

Yeah, it is probably not big enough to be top level in the CNCF. But becoming a subproject under helm may be a good step.

I think the easiest first step toward option 1 would be to figure out how to get the repository moved under the helm governance. so it would be under github.com/helm/helm-diff. Want me to try and see if I can get that going?

> Anyway, thanks again for your feedback! This turned out to be a valuable discussion to build a foundation towards the future of helm-diff.

👍

**Comment by kfox1111** [comment]
Sounds like the next step in the process is outlined here:
https://github.com/helm/community/blob/main/hips/hip-0003.md

I can take the time to file the issue if the project is ok with the requirements listed in the hip.

**Comment by bacongobbler** [comment]
> I think one of the big issues is the limbo state the plugin has because of the copyright issue. If that can be resolved, it may be much more likely that other folks might step in. As, I think it has a chilling effect on contributions.

> What do you mean by this? I see no copyright issue neither in helm-diff nor helm.

I'm also curious what is the copyright issue at-hand here. Both projects are Apache 2.0 licensed, so as long as code authorship attribution rights are respected as per the Apache 2.0 license, there's no issue opening a PR to merge this into Helm's core.

> 1. The governance is moved over to the CNCF. This allows the existing code to be massaged into a form that could be merged directly into helm over time.

I don't understand why that can't be done here. Why does donating the project to the CNCF make it easier to massage the codebase and merge this into helm's core?

The main issue is motivation. Someone must be willing to perform the work involved to open a HIP and go through the review process to merge `helm diff` into helm's core (and find a few Helm maintainers interested in review their code). It's never been a governance issue.

If you're looking to donate the project to Helm (and by extension, the CNCF), that's a different issue entirely.

**Comment by kfox1111** [comment]
> I'm also curious what is the copyright issue at-hand here. Both projects are Apache 2.0 licensed, so as long as code authorship attribution rights are respected as per the Apache 2.0 license, there's no issue opening a PR to merge this into Helm's core.
> 

I clarified that here after I misspoke: https://github.com/databus23/helm-diff/issues/211#issuecomment-1010194877

> > 1. The governance is moved over to the CNCF. This allows the existing code to be massaged into a form that could be merged directly into helm over time.
> 
> I don't understand why that can't be done here. Why does donating the project to the CNCF make it easier to massage the codebase and merge this into helm's core?

See the same issue, around: "This leaves devs to make one of 3 choices:"

> The main issue is motivation. Someone must be willing to perform the work involved to open a HIP and go through the review process to merge `helm diff` into helm's core (and find a few Helm maintainers interested in review their code). It's never been a governance issue.

I'm willing to perform the work involved to open the HIP and go through the process. Been working on that for a while now. :)

> If you're looking to donate the project to Helm (and by extension, the CNCF), that's a different issue entirely.

I think that is part of the process of bringing the helm-diff project under the helm project, or maybe I don't understand the process. Whats the difference between moving under helm's governance and donating it to helm?

I'm looking to be able to either merge the entire project's codebase (likely with modifications) to helm eventually, or the ability to cherry pick code into helm. Governance alignment is a requirement of doing that I think? Anything else?

**Comment by bacongobbler** [comment]
> Whats the difference between moving under helm's governance [...]

Donating the project to the CNCF means the existing maintainers of helm-diff become [Project Maintainers of helm/helm-diff](https://github.com/helm/community/blob/main/governance/governance.md#project-maintainers). The helm/helm Project Maintainers do not automatically become Project Maintainers of helm/helm-diff, though they may ask for the helm/helm-diff maintainers to call for a vote [as per governance](https://github.com/helm/community/blob/main/governance/governance.md#project-maintainers). The CNCF does not magically provide engineering support.

Donating a project to the CNCF grants certain legal protections under the Linux Foundation, as well as assistance in other matters like funding for CI infrastructure, marketing outreach, event planning, etc. The [Helm Org Maintainers](https://github.com/helm/community/blob/main/governance/governance.md#helm-org-maintainers) oversee the project, but the project continues to operate as it did before. [chartmuseum](https://github.com/helm/chartmuseum) and [helm-2to3](https://github.com/helm/helm-2to3) are some examples of community projects that were donated to the CNCF.

The end result to the user is still the same... it's just a different install location: `helm plugin install https://github.com/helm/helm-diff`. I don't think that's what OP (or the rest of the helm-diff community) was asking for when they commented on this ticket... I thought they were asking for `helm diff` to become a new top-level command for Helm.

> [...] and donating it to helm?

When you're referring to "donating it to Helm", I assume you mean "contributing code to Helm". In which case the proecss is the same as you would do when you normally contribute a PR... Which is to [follow the contributing guide](https://github.com/helm/helm/blob/main/CONTRIBUTING.md).

> Governance alignment is a requirement of doing that I think?

Perhaps you're confusing "governance alignment" with "legal due diligence"?

We've merged tons of PRs into Helm without requiring the code authors to align with Helm's governance. They just have to follow the contributing guide and sign off on their work. [Signing your work as per the DCO](https://github.com/helm/helm/blob/main/CONTRIBUTING.md#sign-your-work) certifies that you wrote the patch or otherwise have the right to contribute the material. In this case, you - @kfox1111 - did not write helm-diff in its entirety, so there would be some legal due diligence required as per the Apache 2.0 license agreement. In the past this meant reaching out to the CNCF legal team for advice during the code review process.

**Comment by kfox1111** [comment]
Then as I see the goal, its to donate it to helm? The governance team of helm-diff then stays the same. but the general governance questions that may prevent folks from contributing are all straightened out as part of joining. Contributors trying to work on the goal of refactoring and merging, or cherry picking code from helm-diff to helm then are less encumbered once that is done as the legal due diligence for the existing code is performed as part of the donation and developers doing the legwork only have to worry about their own code rather then the existing codebase?

**Comment by matt314** [comment]
Stumbled onto this and just leaving it here. This is ugly, but does what I need to diff without any additional tool or plugins. Only wrinkle is if you do lookups in helm from the cluster, those do not happen on a dry-run (or template, or any other operation other than install/upgrade). I have default fake values for all my lookups and just know to ignore those in the diff output.
```
% helm upgrade --dry-run pfail-aif-cluster-config aif-cluster-config | grep -A 10000 ^--- | kubectl diff -f -
diff -u -N /var/folders/5w/7sk8s03j1x33903h7ky1rr7w00hgts/T/LIVE-3787786955/operators.coreos.com.v1alpha1.Subscription.openshift-nfd.nfd /var/folders/5w/7sk8s03j1x33903h7ky1rr7w00hgts/T/MERGED-941649389/operators.coreos.com.v1alpha1.Subscription.openshift-nfd.nfd
--- /var/folders/5w/7sk8s03j1x33903h7ky1rr7w00hgts/T/LIVE-3787786955/operators.coreos.com.v1alpha1.Subscription.openshift-nfd.nfd	2022-02-17 12:32:15.000000000 -0500
+++ /var/folders/5w/7sk8s03j1x33903h7ky1rr7w00hgts/T/MERGED-941649389/operators.coreos.com.v1alpha1.Subscription.openshift-nfd.nfd	2022-02-17 12:32:15.000000000 -0500
@@ -5,7 +5,7 @@
     meta.helm.sh/release-name: pfail-aif-cluster-config
     meta.helm.sh/release-namespace: default
   creationTimestamp: "2022-02-15T20:50:31Z"
-  generation: 1
+  generation: 2
   labels:
...
```

**Comment by mumoshu** [comment]
@matt314 Thanks a lot for sharing!

You'll also find it useful to combine `helm upgrade --dry-run` with various flags like `--skip-crds`, `--install`, and `--no-hooks` depending on your use-case. I'd say `--install` is especially important because without that you can't show diff for a new install. Just my two cents.

Also note that `helm-diff` is able to use `helm upgrade --dry-run` instead of `helm template` when you run it with `HELM_DIFF_USE_UPGRADE_DRY_RUN=true` so regardless of whatever backend (helm-template or helm-upgrade-dry-run` ) helm-diff uses, you can simply copy helm-upgrade flags to a helm-diff call to get a diff.

> Only wrinkle is if you do lookups in helm from the cluster,

FYI, this is tracked in https://github.com/databus23/helm-diff/issues/263 and there's on-going PR in upstream https://github.com/helm/helm/pull/9426 to fix that.

> Note, doesn't show changes for resources removed in the Helm charts but still on the cluster. Beware!

Yes. That's why I started using helm-diff years ago and one of the fundamental reasons that I can still justify existence of helm-diff today 😄 

I wonder if `kubectl diff` will eventually get the `--prune` flag that is present in `kubectl apply`, so that it can hopefully detect deleted objects too.

But I think apply's `--prune` has been an alpha feature since years ago and I don't think there's a lot of momentum to complete it and copy it's behavior to kubectl-diff, so...
